src/main/java/com/google/sps/servlets/MealServlet.java

```diff
@@ -177,26 +179,47 @@ private void getMealById(HttpServletRequest request, HttpServletResponse respons
     }
 
     private void returnIdOfSimilar(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        // Fuction redirect to random page
-        // TODO(grenlayk): implement suggestions algorithm here for Product Alpha
+        // Fuction redirect to random page of the same type
         Long mealId = Long.parseLong(getParameter(request, "id", "0"));
-
-        Query query = new Query("Meal");
+        Meal meal = null;
+        Filter idFilter = new FilterPredicate("id", FilterOperator.EQUAL, mealId);
+        Query query = new Query("Meal").setFilter(idFilter);
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        PreparedQuery results = datastore.prepare(query);
-        ArrayList<Long> idList = new ArrayList<>();
-        for (Entity entity : results.asIterable()) {
-            Long id = (Long)entity.getProperty("id");
-            if (id != null && !id.equals(mealId)) {
-                idList.add(id);
-            }
+        List<Meal> meals = getDataFromDatastore(datastore.prepare(query)); 
+        if (meals.size() > 1) {
+            response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+            return;
+        } else if (meals.size() == 1) {
+            meal = (Meal)meals.get(0);
+        } else {
+            response.sendError(HttpServletResponse.SC_NOT_FOUND);
+            return;
         }
+
+        Filter anotherIdFilter = new FilterPredicate("id", FilterOperator.NOT_EQUAL, meal.getId());
+        Filter typeFilter = new FilterPredicate("type", FilterOperator.EQUAL, meal.getType());
+        CompositeFilter sameTypeFilter =
+            CompositeFilterOperator.and(anotherIdFilter, typeFilter);
+        Query typeQuery = new Query("Meal").setFilter(sameTypeFilter);
+        Query idQuery = new Query("Meal").setFilter(anotherIdFilter);
+        List<Meal> sameTypeIdList = getDataFromDatastore(datastore.prepare(typeQuery));
+        List<Meal> idList = getDataFromDatastore(datastore.prepare(idQuery));
+
         Random rand = new Random(); 
-        int index = rand.nextInt(idList.size());
-        Long randomId = idList.get(index);
+        int index = 0;
+        Meal randomMeal = null;
+        if (sameTypeIdList.size() > 0) {
+            index = rand.nextInt(sameTypeIdList.size());
+            randomMeal = (Meal)sameTypeIdList.get(index);
+        } else if (idList.size() > 0) {
+            index = rand.nextInt(idList.size());
+            randomMeal = (Meal)idList.get(index);
```

Comment:
We only use `idList` in this `else` clause, but we fetch the results it in all cases.
We could optimize the logic a bit by delaying evaluation of the `idList` until we actually need it. If you move the `idList = ...` assignment to an `else` block here, you'll avoid an unnecessary request to the database if `sameTypeIdList.size() > 0`.

Reply:
+1 for this idea. By "evaluation of the `idList`" Tomek means to move this entire line over here: `List<Meal> idList = getDataFromDatastore(datastore.prepare(idQuery));` 

In a perfect world your app would never need to retrieve the entire database of entries to work on it. Getting all the entries is expensive both in terms of transferring all the data between Datastore and your app, as well as it may take a lot of RAM to keep all the downloaded data.